### PR TITLE
Handle website update #83

### PR DESCRIPTION
**Pull request by apoorv-2204**
This pr provides api to handle website update. When we upload a website, we might need to do some updates.
Redeploying from scratch is not an option.
Current One seed one website.
We must provide freedom to update website.

we fetch prev ref tx and compare its metadata with the updates files that user has provided.
We update metadata, for the changes files and include new files.

### CODE REVIEW

File: commands/deploy.js [hunk 1]
```diff
@@ -21,34 +22,45 @@ const builder = {
...
+    alias: 'C'
   },
   "ssl-key": {
     describe: 'SSL key to certify the website',
     demandOption: false,
-    type: 'string'
+    type: 'string',
+    alias: 'K'
+  },
+  "website-update": {
```
Comment (samuelmanzanera) [inline commands/deploy.js:55]: I don't think the update should be an option.
It should be the default behavior

**Review by samuelmanzanera — CHANGES_REQUESTED** [review]

File: commands/update.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            // client has added a new file 
+            newFiles.push(pathAsKey);
+        }
+    });
+    return { unchangedMetaData, newFiles };
+}
+
+async function buildLocalRefContent(files) {
+    return files.reduce((acc, { filePath, data }) => {
+        const hash = nativeCryptoLib.createHash('sha-1').update(data).digest('hex')
```
Comment (samuelmanzanera) [inline commands/update.js:53]: 'sha-1' is not supported on node v16, you should us 'sha1'

Comment (samuelmanzanera) [reply commands/update.js:53]: it's throwing an error, you have to be consistent with the lib.js hash function

Comment (apoorv-2204) [reply commands/update.js:53]: Hello the earlier pr has been already merge https://github.com/archethic-foundation/aeweb-cli/issues/94
I have uncommneted try and catch,
please try to use node js version as was in .tools-version in aenode as v18

**Comment by bchamagne** [comment]
I tested 3 scenarios. 2/3 are successful.

- :heavy_check_mark: 1 file changed
- :heavy_check_mark: 2 files changed and 1 directory creation
- :red_circle: 1 directory suppression

The 3rd scenario return "No files to update"

File: commands/deploy.js [hunk 3]
```diff
@@ -90,16 +85,40 @@ const handler = async function(argv) {
...
+      })) {
+        isWebsiteUpdate = true;
+      }
+    }
     // Convert directory structure into array of file content
     console.log(chalk.blue('Creating file structure and compress content...'))
 
     const aeweb = new AEWeb(archethic)
-    const files = cli.getFiles(folderPath, includeGitIgnoredFiles)
+    let files = cli.getFiles(folderPath, includeGitIgnoredFiles)
```
Comment (Neylix) [inline commands/deploy.js:102]: This variable declaration should stay in const as it is never reassigned after

**Review by Neylix — COMMENTED** [review]
Didn't tested yet, put overall looks good, well done !
I have a suggestion, if a user delete files, they remain in the metada and so they are again added in the new reference transaction.
We may list all the files that have been passed throught the `addFile` function, and when we create the reference transaction we can check if there is file in metadata which are not in the file list, so we can delete them from the metadata.

File: commands/deploy.js [hunk 4]
```diff
@@ -90,16 +85,40 @@ const handler = async function(argv) {
     const refIndex = await archethic.transaction.getTransactionIndex(refAddress)
     let filesIndex = await archethic.transaction.getTransactionIndex(filesAddress)
 
+    // Check if website is already deployed
+    if ((await refIndex) !== 0) {
+      console.log(chalk.yellow('Website already deployed'))
+      if (await yesno({
+        question: chalk.yellowBright('Do you want to Update Website Files? (yes/no), no will overwrite the files'
```
[the remark this answers is not in the document]
Comment (apoorv-2204) [reply commands/deploy.js]: But its aeweb cli, options must be there?

**Comment by apoorv-2204** [comment]
> I tested 3 scenarios. 2/3 are successful.
> 
> * ✔️ 1 file changed
> * ✔️ 2 files changed and 1 directory creation
> * 🔴 1 directory deletion
> 
> The 3rd scenario return "No files to update"

Hello. 
I included support for deleted files, so the re tx will be updates if files are deleted.
Directory suppression is not handled as its not a file.

**Comment by samuelmanzanera** [comment]
> > I tested 3 scenarios. 2/3 are successful.
> > 
> > * ✔️ 1 file changed
> > * ✔️ 2 files changed and 1 directory creation
> > * 🔴 1 directory deletion
> > 
> > The 3rd scenario return "No files to update"
> 
> Hello. I included support for deleted files, so the re tx will be updates if files are deleted. Directory suppression is not handled as its not a file.

About the directory deletion, if the previous ref tx reported some files in the deleted directory, you should then remove the files of this given directory (like git is doing)

**Comment by apoorv-2204** [comment]
> > > I tested 3 scenarios. 2/3 are successful.
> > > 
> > > * ✔️ 1 file changed
> > > * ✔️ 2 files changed and 1 directory creation
> > > * 🔴 1 directory deletion
> > > 
> > > The 3rd scenario return "No files to update"
> > 
> > 
> > Hello. I included support for deleted files, so the re tx will be updates if files are deleted. Directory suppression is not handled as its not a file.
> 
> About the directory deletion, if the previous ref tx reported some files in the deleted directory, you should then remove the files of this given directory (like git is doing)

Yes boss we are doing this.

**Comment by apoorv-2204** [comment]
> > > I tested 3 scenarios. 2/3 are successful.
> > > 
> > > * ✔️ 1 file changed
> > > * ✔️ 2 files changed and 1 directory creation
> > > * 🔴 1 directory deletion
> > > 
> > > The 3rd scenario return "No files to update"
> > 
> > 
> > Hello. I included support for deleted files, so the re tx will be updates if files are deleted. Directory suppression is not handled as its not a file.
> 
> About the directory deletion, if the previous ref tx reported some files in the deleted directory, you should then remove the files of this given directory (like git is doing)

I mean, empty directory suppression,  has no meaning, as it has no files. And for files in directory , if they are removed, they are removed from ref tx. Maybe you can give it a try?

File: lib/utils.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        aewebVersion: AEWEB_VERSION,
+        hashFunction: HASH_FUNCTION,
+        metaData: metaData
+    }
+
+    if (sslCertificate) {
+        ref.sslCertificate = sslCertificate
+    }
+
+    return { metadata: metaData, refContent: ref }
```
Comment (Neylix) [inline lib/utils.js:105]: I think you can directly return the stringified content as we expect this function to return the exact content of the transaction

File: lib/api.js [hunk 6]
```diff
@@ -1,41 +1,63 @@
...
-    if (tabPath[0] === '') tabPath.shift()
-    const filePath = tabPath.join('/')
+          // priority given to new files/metadata
+          this.metaData[filePath] = prevFileMetaData;
+
+          // remove from previous metadata
+          delete (this.prevRefTxMetaData[filePath]);
+          return;
+        }
+        else {
```
Comment (Neylix) [inline lib/api.js:47]: I think you can simplify this part like this: 
```javascript
if (this.prevRefTxMetaData !== undefined) {
	// A website update operation
	const prevFileMetaData = this.prevRefTxMetaData[filePath];
	// remove from previous metadata
	delete (this.prevRefTxMetaData[filePath]);

	if (prevFileMetaData != undefined && prevFileMetaData["hash"] == hash) {
		// File already exists
		this.metaData[filePath] = prevFileMetaData;
		return;
	}
}
// File has changed or is new
this.modifiedFiles.push(filePath);
```

**Review by Neylix — CHANGES_REQUESTED** [review]

File: lib/api.js [hunk 7]
```diff
@@ -62,15 +71,16 @@ export default class AEWeb {
 
       return tx
     })
+
   }
 
   async getRefTransaction(transactions) {
-    const { metaData, refContent } = getMetaData(this.txsContent, transactions, this.metaData, this.sslCertificate)
+    const { metaData, refContent } = getRefTxContent(this.txsContent, transactions, this.metaData, this.sslCertificate)
```
Comment (Neylix) [inline lib/api.js:78]: There it will crash since the returned value of getRefTxContent is a string and not anymore an object

**Review by Neylix — CHANGES_REQUESTED** [review]

File: lib/api.js (hunk 7 above)
Comment (apoorv-2204) [reply lib/api.js]: thanks.tested and rebased

**Review by Neylix — APPROVED** [review]

**Pull request merged**